File: 061_rotate_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return head;

        ListNode curr = head;
        int len = 1;
        while (curr.next != null) {
            curr = curr.next;
            len++;
        }
        if (len == 1) return head;

        int offset = len - (k % len);

        ListNode newHead = head;
        ListNode newHeadBefOne = null;
        while (offset > 0) {
            if (offset == 1) newHeadBefOne = newHead;
            newHead = newHead.next;
            if (newHead == null) newHead = head;
            offset--;
        }

        if (newHead != head){ // then append remaining to newHead
            ListNode node = newHead;
            while (node.next != null) node = node.next;
            node.next = head; // append from the first
            newHeadBefOne.next = null; // cut off the remaining
        }
        return newHead;
    }
}

File: python/061_rotate_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # form a cycle and move (count - k) steps, and create a new head
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or k == 0: return head
        tail = head
        count = 1
        while tail.next:
            tail = tail.next
            count += 1
        tail.next = head
        k %= count
        offset = count - k
        for _ in range(offset):
            tail = tail.next
            offset -= 1
        head = tail.next
        tail.next = None
        return head

    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head: return head
        leng = 0
        curr = head
        while curr:
            curr = curr.next
            leng += 1
        if leng == 1:
            return head

        k %= leng
        offset = leng - k

        curr = head
        prev = None
        while offset > 0:
            if offset == 1:
                prev = curr
            curr = curr.next
            if not curr:
                curr = head
            offset -= 1

        if curr != head:
            tail = curr
            while tail.next:
                tail = tail.next
            prev.next = None
            tail.next = head

        return curr


